File: leetcode/1503/solution.go
/**
 * https://leetcode.com/problems/last-moment-before-all-ants-fall-out-of-a-plank/
 * 1503. Last Moment Before All Ants Fall Out of a Plank
 */

package main

func getLastMoment(n int, left []int, right []int) int {
	longest_l := max(left)
	longest_r := n - min(right)
	if longest_l > longest_r {
		return longest_l
	}
	return longest_r
}

func max(args []int) int {
	if len(args) == 0 {
		return 0
	}
	a := args[0]
	for _, b := range args {
		if a < b {
			a = b
		}
	}
	return a
}

func min(args []int) int {
	if len(args) == 0 {
		return 10001 // maximum value
	}
	a := args[0]
	for _, b := range args {
		if a > b {
			a = b
		}
	}
	return a
}

File: leetcode/1503/solution.rs
/**
 * https://leetcode.com/problems/last-moment-before-all-ants-fall-out-of-a-plank/
 * 1503. Last Moment Before All Ants Fall Out of a Plank
 */

impl Solution {
    pub fn get_last_moment(n: i32, left: Vec<i32>, right: Vec<i32>) -> i32 {
        match (left.iter().max(), right.iter().min()) {
            (None, None) => 0,
            (None, Some(&r)) => n - r,
            (Some(&l), None) => l,
            (Some(&l), Some(&r)) => l.max(n - r),
        }
    }
}
